core/src/test/java/com/linecorp/armeria/client/ClientGoAwayGracefulShutdownTest.java

```diff
@@ -0,0 +1,96 @@
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.server.logging.LoggingService;
+import com.linecorp.armeria.testing.junit5.server.ServerExtension;
+
+import io.netty.util.AttributeMap;
+
+class ClientGoAwayGracefulShutdownTest {
+    private static final AtomicInteger counter = new AtomicInteger();
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.service("/foo", (ctx, req) -> {
+                if (counter.getAndIncrement() == 0) {
+                    return HttpResponse.of("OK");
+                } else {
+                    return HttpResponse.delayed(HttpResponse.of("OK"), Duration.ofSeconds(5));
+                }
+            });
+            // 1. A GOAWAY with Integer.MAX_VALUE will be sent in 1 second.
+            sb.idleTimeout(Duration.ofSeconds(1));
+            // 2. A GOAWAY with the last known stream identifier(3) in 2 seconds after the graceful period.
+            sb.connectionDrainDuration(Duration.ofSeconds(2));
+            sb.decorator(LoggingService.newDecorator());
+        }
+    };
+
+    @Test
+    void gracefulShutdown() throws InterruptedException {
+        final AtomicInteger opened = new AtomicInteger();
+        final ConnectionPoolListener poolListener = new ConnectionPoolListener() {
+            @Override
+            public void connectionOpen(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                       InetSocketAddress localAddr, AttributeMap attrs) throws Exception {
+                opened.incrementAndGet();
+            }
+
+            @Override
+            public void connectionClosed(SessionProtocol protocol, InetSocketAddress remoteAddr,
+                                         InetSocketAddress localAddr, AttributeMap attrs) throws Exception {}
+        };
+
+        try (ClientFactory factory = ClientFactory.builder()
+                                                  .idleTimeoutMillis(0)
+                                                  .connectionPoolListener(poolListener)
+                                                  .build()) {
+
+            final WebClient client = WebClient.builder(server.httpUri())
+                                              .factory(factory)
+                                              .decorator(LoggingClient.newDecorator())
+                                              .responseTimeoutMillis(0)
+                                              .build();
+            client.get("/foo").aggregate().join();
+            // Trigger the idle timeout and wait for a connection to initiate the graceful shutdown.
+            Thread.sleep(2000);
+            assertThat(opened).hasValue(1);
+
+            // A client should open another connection and
+            // should not send a request using the closing connection.
```

Comment:
How about checking the value of `opened` to confirm this statement?

Reply:
Fixed.